File: engines/typescript/src/canonical.ts
/**
 * The AgentCE canonical form (SPEC §6.7): RFC 8785 JCS under the evidence profile.
 *
 * The evidence profile forbids non-integer JSON numbers, so the hard part of JCS — shortest
 * round-trip formatting of IEEE-754 doubles — never arises, and the canonical form is a small exact
 * set of rules every engine reproduces byte for byte (see `spec/rules`/`canonical-form.md`, the
 * shared vectors under `spec/model/test-vectors/`). This is the TypeScript engine's own
 * implementation; it must agree with the Python reference to the byte.
 *
 * A JSON number is accepted only when its token is an integer (no fraction, no exponent) of at most
 * 2^53 - 1 in magnitude. `1.5`, `0.1`, `1e+30`, `1.0`, `1e2` and `-0.0` are refused as
 * `non_integer_number`, and `9007199254740992` and above as `integer_out_of_range`, exactly as the
 * reference does. `JSON.parse` folds those tokens to ordinary numbers before the canonical form can
 * see them, so evidence is read with `parseJson` (`./json`), which keeps them refusable.
 * No network, no learned component.
 */

import { createHash } from "node:crypto";

import { NonCanonicalNumber } from "./json";

export class CanonicalizationError extends Error {
  readonly reason: string;

  constructor(reason: string, detail = "") {
    super(detail ? `${reason}: ${detail}` : reason);
    this.reason = reason;
    this.name = "CanonicalizationError";
  }
}

const SHORT_ESCAPES = new Map<number, string>([
  [0x08, "\\b"],
  [0x09, "\\t"],
  [0x0a, "\\n"],
  [0x0c, "\\f"],
  [0x0d, "\\r"],
  [0x22, '\\"'],
  [0x5c, "\\\\"],
]);

function nfc(text: string): string {
  return text.normalize("NFC");
}

function serialiseString(text: string): string {
  const out: string[] = ['"'];
  for (const ch of nfc(text)) {
    const code = ch.codePointAt(0) as number;
    const escaped = SHORT_ESCAPES.get(code);
    if (escaped !== undefined) {
      out.push(escaped);
    } else if (code < 0x20) {
      out.push(`\\u${code.toString(16).padStart(4, "0")}`);
    } else {
      out.push(ch);
    }
  }
  out.push('"');
  return out.join("");
}

function serialiseObject(obj: Record<string, unknown>): string {
  const members: Array<[string, unknown]> = [];
  const seen = new Set<string>();
  for (const rawKey of Object.keys(obj)) {
    const key = nfc(rawKey);
    if (seen.has(key)) {
      throw new CanonicalizationError("duplicate_key_after_nfc", key);
    }
    seen.add(key);
    members.push([key, obj[rawKey]]);
  }
  // JS string comparison is by UTF-16 code unit, matching the reference's UTF-16-BE byte order.
  members.sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0));
  return `{${members.map(([key, val]) => `${serialiseString(key)}:${serialise(val)}`).join(",")}}`;
}

function serialise(value: unknown): string {
  if (value === null) {
    return "null";
  }
  if (typeof value === "boolean") {
    return value ? "true" : "false";
  }
  if (value instanceof NonCanonicalNumber) {
    throw new CanonicalizationError(value.reason, value.source);
  }
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) {
      const reason = Number.isInteger(value) ? "integer_out_of_range" : "non_integer_number";
      throw new CanonicalizationError(reason, String(value));
    }
    return String(value);
  }
  if (typeof value === "string") {
    return serialiseString(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map(serialise).join(",")}]`;
  }
  if (typeof value === "object") {
    return serialiseObject(value as Record<string, unknown>);
  }
  throw new CanonicalizationError("unsupported_type", typeof value);
}

/** The canonical form of a JSON value as a string (SPEC §6.7). */
export function canonicalString(value: unknown): string {
  return serialise(value);
}

/** The canonical form of a JSON value as UTF-8 bytes (SPEC §6.7). */
export function canonicalize(value: unknown): Buffer {
  return Buffer.from(canonicalString(value), "utf-8");
}

/** The lowercase hex SHA-256 of the canonical form (basis of `integrity.hash`, SPEC §6.6). */
export function sha256Hex(value: unknown): string {
  return createHash("sha256").update(canonicalize(value)).digest("hex");
}

File: engines/java/src/main/java/org/agentce/Canonical.java
package org.agentce;

import com.fasterxml.jackson.databind.JsonNode;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.text.Normalizer;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * The AgentCE canonical form (SPEC §6.7): RFC 8785 JCS under the evidence profile.
 *
 * <p>The evidence profile forbids non-integer JSON numbers, so the hard part of JCS — shortest
 * round-trip formatting of IEEE-754 doubles — never arises, and the canonical form is a small exact
 * set of rules every engine reproduces byte for byte (see {@code spec/rules/canonical-form.md} and the
 * shared vectors under {@code spec/model/test-vectors/}). This is the Java engine's own implementation;
 * it must agree with the Python reference and the TypeScript engine to the byte.
 *
 * <p>A JSON number is an integer exactly when its source token carries no {@code .}, {@code e}, or
 * {@code E} — the same rule the reference's {@code json.loads} applies (an {@code int} versus a
 * {@code float}). Jackson exposes that distinction as {@link JsonNode#isIntegralNumber()}, so {@code
 * 1e+30}, {@code 1.0}, {@code 1e2} and {@code 0.1} are refused as {@code non_integer_number}, an
 * integer beyond 2^53 - 1 in magnitude as {@code integer_out_of_range}, and {@code 9007199254740991}
 * is kept. No network, no learned component.
 */
public final class Canonical {
    private Canonical() {}

    /** The range every engine represents exactly: an integer beyond 2^53 - 1 in magnitude is refused. */
    private static final BigInteger MAX_SAFE_INTEGER = BigInteger.ONE.shiftLeft(53).subtract(BigInteger.ONE);

    /** A value cannot be put in canonical form under the AgentCE profile (stable {@code reason}). */
    public static final class CanonicalizationError extends RuntimeException {
        private static final long serialVersionUID = 1L;
        public final String reason;

        public CanonicalizationError(String reason, String detail) {
            super(detail == null || detail.isEmpty() ? reason : reason + ": " + detail);
            this.reason = reason;
        }
    }

    private static String shortEscape(int code) {
        return switch (code) {
            case 0x08 -> "\\b";
            case 0x09 -> "\\t";
            case 0x0A -> "\\n";
            case 0x0C -> "\\f";
            case 0x0D -> "\\r";
            case 0x22 -> "\\\"";
            case 0x5C -> "\\\\";
            default -> null;
        };
    }

    private static String nfc(String text) {
        return Normalizer.normalize(text, Normalizer.Form.NFC);
    }

    private static void serialiseString(String text, StringBuilder out) {
        out.append('"');
        String normalised = nfc(text);
        // Iterate UTF-16 code units: control units are escaped; surrogates (>= 0xD800) are appended
        // verbatim and encode to their correct UTF-8 bytes when the string is written out.
        for (int i = 0; i < normalised.length(); i++) {
            char ch = normalised.charAt(i);
            int code = ch;
            String escape = shortEscape(code);
            if (escape != null) {
                out.append(escape);
            } else if (code < 0x20) {
                out.append(String.format("\\u%04x", code));
            } else {
                out.append(ch);
            }
        }
        out.append('"');
    }

    private static void serialiseObject(JsonNode node, StringBuilder out) {
        List<String> keys = new ArrayList<>();
        List<JsonNode> values = new ArrayList<>();
        Set<String> seen = new HashSet<>();
        var fields = node.fieldNames();
        // Field iteration is insertion order; two keys equal only after NFC are the ambiguous case.
        var it = node.fields();
        while (it.hasNext()) {
            var entry = it.next();
            String key = nfc(entry.getKey());
            if (!seen.add(key)) {
                throw new CanonicalizationError("duplicate_key_after_nfc", key);
            }
            keys.add(key);
            values.add(entry.getValue());
        }
        // Sort by UTF-16 code unit (String.compareTo), matching the reference's utf-16-be byte order.
        Integer[] order = new Integer[keys.size()];
        for (int i = 0; i < order.length; i++) {
            order[i] = i;
        }
        java.util.Arrays.sort(order, (a, b) -> keys.get(a).compareTo(keys.get(b)));
        out.append('{');
        for (int i = 0; i < order.length; i++) {
            if (i > 0) {
                out.append(',');
            }
            serialiseString(keys.get(order[i]), out);
            out.append(':');
            serialise(values.get(order[i]), out);
        }
        out.append('}');
        // fields is unused beyond documenting insertion order; keep the reference readable.
        assert fields != null;
    }

    private static void serialise(JsonNode value, StringBuilder out) {
        if (value == null || value.isNull()) {
            out.append("null");
        } else if (value.isBoolean()) {
            out.append(value.booleanValue() ? "true" : "false");
        } else if (value.isNumber()) {
            if (value.isIntegralNumber()) {
                BigInteger integer = value.bigIntegerValue();
                if (integer.abs().compareTo(MAX_SAFE_INTEGER) > 0) {
                    throw new CanonicalizationError("integer_out_of_range", integer.toString());
                }
                out.append(integer.toString());
            } else {
                throw new CanonicalizationError("non_integer_number", value.asText());
            }
        } else if (value.isTextual()) {
            serialiseString(value.textValue(), out);
        } else if (value.isArray()) {
            out.append('[');
            for (int i = 0; i < value.size(); i++) {
                if (i > 0) {
                    out.append(',');
                }
                serialise(value.get(i), out);
            }
            out.append(']');
        } else if (value.isObject()) {
            serialiseObject(value, out);
        } else {
            throw new CanonicalizationError("unsupported_type", value.getNodeType().toString());
        }
    }

    /** The canonical form of a JSON value as a string (SPEC §6.7). */
    public static String canonicalString(JsonNode value) {
        StringBuilder out = new StringBuilder();
        serialise(value, out);
        return out.toString();
    }

    /** The canonical form of a JSON value as UTF-8 bytes (SPEC §6.7). */
    public static byte[] canonicalize(JsonNode value) {
        return canonicalString(value).getBytes(StandardCharsets.UTF_8);
    }

    /** The lowercase hex SHA-256 of the canonical form (basis of {@code integrity.hash}, SPEC §6.6). */
    public static String sha256Hex(JsonNode value) {
        return sha256Hex(canonicalize(value));
    }

    /** The lowercase hex SHA-256 of arbitrary bytes. */
    public static String sha256Hex(byte[] data) {
        try {
            byte[] digest = MessageDigest.getInstance("SHA-256").digest(data);
            StringBuilder hex = new StringBuilder(digest.length * 2);
            for (byte b : digest) {
                hex.append(Character.forDigit((b >> 4) & 0xF, 16));
                hex.append(Character.forDigit(b & 0xF, 16));
            }
            return hex.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 unavailable", e);
        }
    }
}
